### PR TITLE
BUG: assignment to multiple columns when some column do not exist

### LINKED ISSUE #13658: Assignment to multiple columns only works if they existed before
```
import pandas as pd
df = pd.DataFrame({'a': [1, 2]})
df['b'] = 3 # creates column 'b'
df[['a', 'b']] = 4 # assigns to columns 'a' and 'b'
df[['c', 'd']] = 5 # KeyError: "['c' 'd'] not in index" !
df[['b', 'c']] = 6 # KeyError: "['c'] not in index" !
```

I would have expected all the above cases to work, but for some reason the last two fail.  New column creation only seems to work for a single column, whilst multiple-column assignment works only if all columns exist.

> INSTALLED VERSIONS
> python: 3.5.1.final.0
> python-bits: 64
> pandas: 0.18.1
> numpy: 1.11.1

### PR DESCRIPTION
**Pull request by howsiwei**
- [x] closes #13658
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

In particular, the following code now behaves correctly.

```python
import pandas as pd
df = pd.DataFrame({'a': [0, 1, 2], 'b': [3, 4, 5]})
df[['a', 'c']] = 1
print(df)
```
v0.22: error

master: column `'c'` is converted to index `-1`, which causes last column to be overwritten.
```
   a  b
0  1  1
1  1  1
2  1  1
```

After this PR:
```
   a  b  c
0  1  3  1
1  1  4  1
2  1  5  1
```

### CODE REVIEW

**Comment by howsiwei** [comment]
@jreback any feedback?

File: pandas/tests/frame/test_indexing.py [hunk 1]
```diff
@@ -185,6 +186,49 @@ def test_setitem_list_of_tuples(self):
         expected = Series(tuples, index=self.frame.index, name='tuples')
         assert_series_equal(result, expected)
 
+    def test_setitem_list_missing_columns(self):
```
Comment (simonjayhawkins) [inline pandas/tests/frame/test_indexing.py]: can you split this test so that the two cases highlighted in the issue are separate tests. (or parametrize if more appropriate) and add the issue number as a comment.

**Review by simonjayhawkins — COMMENTED** [review]
@howsiwei Thanks for the PR.

IIUC the issue relates to the scalar assignment to multiple columns of a DataFrame using __setitem__ only.

test_loc_setitem_corner in pandas\tests\series\indexing\test_loc.py is currently failing. Changing the behavior of Series assignment is not covered in the scope of the issue.

File: doc/source/whatsnew/v0.25.0.rst [hunk 2]
```diff
@@ -425,6 +425,7 @@ Indexing
 - Bug in which :meth:`DataFrame.to_csv` caused a segfault for a reindexed data frame, when the indices were single-level :class:`MultiIndex` (:issue:`26303`).
 - Fixed bug where assigning a :class:`arrays.PandasArray` to a :class:`pandas.core.frame.DataFrame` would raise error (:issue:`26390`)
 - Allow keyword arguments for callable local reference used in the :method:`DataFrame.query` string (:issue:`26426`)
+- Bug in assignment to multiple columns of a `DataFrame` when some of the columns do not exist (:issue:`13658`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: use ``:class:`DataFrame` ``; this is assignment with a scalar, yes?

**Review by jreback — CHANGES_REQUESTED** [review]
this is adding quite a bit of code; please try to streamline and not special case

File: doc/source/whatsnew/v0.25.0.rst (hunk 2 above)
Comment (howsiwei) [reply doc/source/whatsnew/v0.25.0.rst]: The assignment should allow the same set of values as when the columns are present.

Comment (howsiwei) [reply doc/source/whatsnew/v0.25.0.rst]: Hi, thanks for the feedback. Any pointer on how to streamline the code?

**Comment by howsiwei** [comment]
Hi, thanks for the feedback. Any pointer on how to streamline the code?

File: pandas/tests/frame/test_indexing.py (hunk 1 above)
Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: Updated. Thanks for the feedback.

File: pandas/tests/frame/test_indexing.py [hunk 3]
```diff
@@ -185,6 +186,51 @@ def test_setitem_list_of_tuples(self):
...
+        expected = self.frame.columns.union(['E'])
+        assert_index_equal(result, expected)
+
+        result = df[self.frame.columns.drop('A')]
+        expected = self.frame.drop('A', axis=1)
+        assert_frame_equal(result, expected)
+
+        result = df[['A', 'E']]
+        expected = self.frame[['B', 'C']]
+        assert_numpy_array_equal(result.to_numpy(), expected.to_numpy())
```
Comment (simonjayhawkins) [inline pandas/tests/frame/test_indexing.py]: i'm having a hard time just eyeballing these tests. Ideally, should have just one assert per test. construct the expected and use `assert_frame_equal` which checks indexes, values, labels, names etc. with one assert. 

Do let me know if i've misunderstood the intention here.

Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: The problem is how to obtain the result `DataFrame`. I suppose one way to do that is to assume the correctness of assigning a single column and obtain the result by assigning a single column multiple times.

Comment (simonjayhawkins) [reply pandas/tests/frame/test_indexing.py]: yes. that sounds good.

Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: Updated.

File: doc/source/whatsnew/v0.25.0.rst (hunk 2 above)
Comment (simonjayhawkins) [reply doc/source/whatsnew/v0.25.0.rst]: > use `` :class:`DataFrame` ``

can you update.

File: pandas/tests/frame/test_indexing.py [hunk 4]
```diff
@@ -185,6 +186,36 @@ def test_setitem_list_of_tuples(self):
         expected = Series(tuples, index=self.frame.index, name='tuples')
         assert_series_equal(result, expected)
 
+    def test_setitem_list_all_missing_columns_scalar(self):
+        df = self.frame.copy()
```
Comment (simonjayhawkins) [inline pandas/tests/frame/test_indexing.py]: use result =

File: pandas/tests/frame/test_indexing.py [hunk 5]
```diff
@@ -185,6 +186,36 @@ def test_setitem_list_of_tuples(self):
...
+        df[['E', 'F']] = 1
+
+        df2 = self.frame.copy()
```
Comment (simonjayhawkins) [inline pandas/tests/frame/test_indexing.py]: use expected =

File: pandas/tests/frame/test_indexing.py [hunk 6]
```diff
@@ -185,6 +186,36 @@ def test_setitem_list_of_tuples(self):
...
+        df2['E'] = 1
+        df2['F'] = 1
+
+        assert_frame_equal(df, df2)
+
+    def test_setitem_list_some_missing_columns_list(self):
+        df = self.frame.copy()
```
Comment (simonjayhawkins) [inline pandas/tests/frame/test_indexing.py]: use float_frame fixture (if you can)

File: pandas/tests/frame/test_indexing.py [hunk 7]
```diff
@@ -185,6 +186,36 @@ def test_setitem_list_of_tuples(self):
...
+        df[['A', 'E']] = [1, 2]
+
+        df2 = self.frame.copy()
+        df2['A'] = 1
+        df2['E'] = 2
+
+        assert_frame_equal(df, df2)
+
+    def test_setitem_list_some_missing_columns_dataframe(self):
```
Comment (simonjayhawkins) [inline pandas/tests/frame/test_indexing.py]: add GH issue number

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
tests are much easier to read. thanks.

a few minor comments on the tests

we are also going to need another test for scalar assignment with some missing columns to cover the other issue.  unless we have one.

do we have a test to check these cases raise a warning on getting after this change?

I think we should also have a test to now check that the message is not raised when setting (or do we have a test for this?) 

not looked at the actual code, too many tests are currently failing.

File: doc/source/whatsnew/v0.25.0.rst (hunk 2 above)
Comment (howsiwei) [reply doc/source/whatsnew/v0.25.0.rst]: Updated

File: pandas/tests/frame/test_indexing.py (hunk 4 above)
Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: Updated

File: pandas/tests/frame/test_indexing.py (hunk 5 above)
Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: Updated

File: pandas/tests/frame/test_indexing.py (hunk 6 above)
Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: What's float_frame fixture? Mind giving me some pointer?

File: pandas/tests/frame/test_indexing.py (hunk 7 above)
Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: Updated

File: pandas/core/indexing.py [hunk 8]
```diff
@@ -462,14 +471,20 @@ def _setitem_with_indexer(self, indexer, value):
             if isinstance(value, ABCSeries):
                 value = self._align_series(indexer, value)
 
-            info_idx = indexer[info_axis]
-            if is_integer(info_idx):
-                info_idx = [info_idx]
-            labels = item_labels[info_idx]
+            if has_missing_columns:
+                labels = [idx['key'] if isinstance(idx, dict) else
+                          item_labels[idx]
```
Comment (jreback) [inline pandas/core/indexing.py]: @howsiwei you are adding WAY too much complexity here. You need to see the minimal change set to make this change work, it is likely way simpler.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 8 above)
Comment (jreback) [reply pandas/core/indexing.py]: trace other setter calls and see

Comment (howsiwei) [reply pandas/core/indexing.py]: @jreback mind suggesting where I should make the changes? The code for setting values is quite complicated and it's not obvious to me at all how to achieve the desired result with fewer changes.

File: pandas/tests/frame/test_indexing.py (hunk 6 above)
Comment (simonjayhawkins) [reply pandas/tests/frame/test_indexing.py]: have a look at #25633.

the other tests in this module are being updated to use the float_frame fixture.

Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: Updated.

File: pandas/core/indexing.py (hunk 8 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: @jreback since `value` can have multiple types (eg. scalar, 1d array, 2d array, `DataFrame`), it's not easy to split `value` into value for each column. The only way I see to not duplicate the value splitting code is to modify the helper function `setter`, where `value` has already been split. I don't see how the code can be made much simpler when using this way.

Comment (howsiwei) [reply pandas/core/indexing.py]: @jreback any updates?

Comment (howsiwei) [reply pandas/core/indexing.py]: @jreback any suggestions?

Comment (jreback) [reply pandas/core/indexing.py]: no suggestions as I have not looked at this recently
next week possibly

Comment (howsiwei) [reply pandas/core/indexing.py]: @jreback sorry to disturb you once again. Have you looked into this PR?

Comment (jreback) [reply pandas/core/indexing.py]: @howsiwei honestly i have not.

File: pandas/tests/frame/test_indexing.py [hunk 9]
```diff
@@ -1093,10 +1127,11 @@ def test_fancy_index_int_labels_exceptions(self):
                    r" dtype='object'\)\] are in the \[index\]")
             with pytest.raises(KeyError, match=msg):
                 self.frame.ix[['foo', 'bar', 'baz']] = 1
-            msg = (r"None of \[Index\(\['E'\], dtype='object'\)\] are in the"
-                   r" \[columns\]")
-            with pytest.raises(KeyError, match=msg):
-                self.frame.ix[:, ['E']] = 1
+            # partial setting now allows this GH13658
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: don;t comment out, simply remove

File: pandas/core/indexing.py (hunk 8 above)
Comment (jreback) [reply pandas/core/indexing.py]: so this is what I would do. I would touch no code here at all; rather add what you need into the loop above around 328. This handles the missing indexer case. 

Then we can assert that we have *no* missing items at all, IOW what you have at line 307 (what you define has_missing columns), could be true there but MUST be false at line 465.

This entire setting routine is *very* complex. We want to make this simpler and easier to understand; that may take a bigger refactor, but let's start by not adding additional complexity as much as possible.

i know this is non-trivial, but I am holding the line here because this is *already* pretty impenetrable code.

love for you to make it better!

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 8 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: @jreback thanks for the suggestion. I'm guessing what you mean is that we could simply insert the column before setting. There are 2 downsides to this approach though:
- It may be less efficient.
- More importantly, currently `pandas` would try to infer the fill values for a column. Consider the following code:

``` python
import numpy as np
import pandas as pd

df = pd.DataFrame({'a': [0, 0]})
df.loc[0, 'b'] = pd.Timestamp('20120101')
df.loc[0, 'c'] = 1.0
print(df)
```

Output:
```
   a                    b    c
0  0  2012-01-01 00:00:00  1.0
1  0                  NaT  NaN
```

Note that column `b` and `c` are filled with `NaT` and `NaN` respectively.

To infer the fill values for each column, we need to split the values into separated columns first. But this splitting is only done after line 465.

File: pandas/tests/frame/test_indexing.py (hunk 9 above)
Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: I commented out the test following the test immediately below on line 1136. Should it be removed too?

File: pandas/core/indexing.py (hunk 8 above)
Comment (jreback) [reply pandas/core/indexing.py]: we already do this for expansions

my point is you are introducing another code path here

so either remove and fix the existing one (non trivial) or integrate this change

the dtype changes are expected in the current implementation

Comment (howsiwei) [reply pandas/core/indexing.py]: > we already do this for expansions

Do what?

> the dtype changes are expected in the current implementation

What do you mean?

Comment (howsiwei) [reply pandas/core/indexing.py]: @jreback Hmm after reading this again I think I understand what you said. Do you mean that using `NaN` as fill values for all columns is fine because this is already done elsewhere in pandas code? What are expansions though?

Comment (howsiwei) [reply pandas/core/indexing.py]: @jreback can you please confirm if I should just fix this issue by inserting columns filled with `NaN` first?

File: pandas/tests/frame/test_indexing.py (hunk 9 above)
Comment (jreback) [reply pandas/tests/frame/test_indexing.py]: yes try that

Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: > yes try that

Yes for this question too?

> @jreback can you please confirm if I should just fix this issue by inserting columns filled with `NaN` first?

**Comment by howsiwei** [comment]
@jreback what do you think about the current changes? Also do you have any idea why It's failing on Linux py37_np_dev?

File: pandas/core/indexing.py [hunk 10]
```diff
@@ -1266,9 +1278,23 @@ def _convert_to_indexer(
                 inds, = obj.nonzero()
                 return inds
             else:
-                # When setting, missing keys are not allowed, even with .loc:
-                kwargs = {"raise_missing": True if is_setter else raise_missing}
-                return self._get_listlike_indexer(obj, axis, **kwargs)[1]
+                try:
+                    # When setting, missing keys are not allowed, even with # .loc:
+                    kwargs = {"raise_missing": True if is_setter else raise_missing}
```
Comment (jreback) [inline pandas/core/indexing.py]: you don't want to do this in a try/except as its extremely error prone. Instead if you have a is_setter you can just make sure the columns exist (but doing the _convert_to_indexer).

I think if you do this *first* e.g. almost first in _setitem_with_indexer you won't need to make the code change you have above about taking the split path.

File: pandas/tests/frame/test_indexing.py [hunk 11]
```diff
@@ -208,6 +208,43 @@ def test_setitem_list_of_tuples(self, float_frame):
         expected = Series(tuples, index=float_frame.index, name="tuples")
         assert_series_equal(result, expected)
 
+    def test_setitem_list_all_missing_columns_scalar(self, float_frame):
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: can you parameterize these, typicall we add an arg called box to do this, eg.

```
@pytest.mark.parameterize('box', [lambda x: 1, lambda x: [1, 2], lambda x: x[['B', 'C']]])
...
```

File: pandas/tests/indexing/test_loc.py [hunk 12]
```diff
@@ -808,6 +808,46 @@ def test_loc_setitem_with_scalar_index(self, indexer, value):
 
         assert is_scalar(result) and result == "Z"
 
+    def test_loc_setitem_missing_columns_scalar_index_list_value(self):
```
Comment (jreback) [inline pandas/tests/indexing/test_loc.py]: same as above

File: doc/source/whatsnew/v0.25.1.rst [hunk 13]
```diff
@@ -85,7 +85,7 @@ Interval
 Indexing
 ^^^^^^^^
 
--
+- Bug in assignment to multiple columns of a :class:`DataFrame` when some of the columns do not exist (:issue:`13658`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.1.rst]: this will need a subsection to explain the change. move to 1.0 as this is a non-trivial change of a longstanding issue.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 10 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: I force `take_split_path = True` to work around #27583. Otherwise I would fail tests such as `test_setitem_list_all_missing_columns_scalar` in  `pandas/tests/frame/test_indexing.py`.

File: pandas/tests/frame/test_indexing.py (hunk 11 above)
Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: I can parametrize the right-hand size of an assignment but how do I parametrize the left-hand size?

File: doc/source/whatsnew/v0.25.1.rst (hunk 13 above)
Comment (howsiwei) [reply doc/source/whatsnew/v0.25.1.rst]: I have added a subsection but the docs is failing. Any pointers on what should I change?

**Comment by howsiwei** [comment]
@jreback I have moved my entire fix to earlier stage of setting as https://github.com/pandas-dev/pandas/pull/27604 remove `is_setter` which is required in my earlier approach. What do you think about it?

File: pandas/core/frame.py [hunk 14]
```diff
@@ -3007,6 +3007,12 @@ def _setitem_array(self, key, value):
                 for k1, k2 in zip(key, value.columns):
                     self[k1] = value[k2]
             else:
+                if all(is_hashable(k) for k in key):
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Under what cases can elements of `key` not be hashable? These are our eventual column names, right? Which I *think* we require they be hashable.

I'm a bit concerned about the performance here, for the case when `key` is large (which is possible, right?)

**Review by TomAugspurger — COMMENTED** [review]
Jeff may be offline for a bit. We'll get to this before 1.0 though.

File: pandas/core/frame.py (hunk 14 above)
Comment (howsiwei) [reply pandas/core/frame.py]: The only reason that I check whether `key` is hashable is to produce the correct? errors, which is tested in https://github.com/pandas-dev/pandas/blob/e55b6980fd5c99087d0959952b22e74434e9f5e2/pandas/tests/indexing/test_indexing.py#L180-L184
If producing all these errors are not required then this check can be omitted.

Comment (howsiwei) [reply pandas/core/frame.py]: @TomAugspurger what's your thought on this?

Comment (TomAugspurger) [reply pandas/core/frame.py]: What happens if the check is removed?

Comment (howsiwei) [reply pandas/core/frame.py]: The error produced is different.

File: doc/source/whatsnew/v1.0.0.rst [hunk 15]
```diff
@@ -40,6 +40,37 @@ Backwards incompatible API changes
 - :class:`pandas.core.groupby.GroupBy.transform` now raises on invalid operation names (:issue:`27489`).
 -
 
+.. _whatsnew_1000.api_breaking.multicolumn_assignment:
+
+Assignment to multiple columns of a DataFrame when some columns do not exist
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Assignment to multiple columns of a :class:`DataFrame` when some of the columns do not exist would previously assign the values to the last column. Now, new columns would be constructed withe the right values. (:issue:`13658`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.0.0.rst]:
```suggestion
Assignment to multiple columns of a :class:`DataFrame` when some of the columns do not exist would previously assign the values to the last column. Now, new columns would be constructed with the right values. (:issue:`13658`)
```

File: pandas/core/frame.py (hunk 14 above)
Comment (WillAyd) [reply pandas/core/frame.py]: Yea this seems strange to me. I feel like this *should* be hashable inherently so clarification here would be helpful

File: pandas/core/indexing.py [hunk 16]
```diff
@@ -197,6 +198,19 @@ def _get_setitem_indexer(self, key):
     def __setitem__(self, key, value):
         if isinstance(key, tuple):
             key = tuple(com.apply_if_callable(x, self.obj) for x in key)
+            if (
```
Comment (WillAyd) [inline pandas/core/indexing.py]: Maybe this gets simplified if the `hashable` stuff can be removed, but can you add a comment here as to what is going on? Not immediately apparent what all of these conditions are

File: pandas/tests/frame/test_indexing.py [hunk 17]
```diff
@@ -208,6 +208,29 @@ def test_setitem_list_of_tuples(self, float_frame):
...
+            lambda x: np.array([1, 2]),
+            lambda x: x[["B", "C"]],
+            lambda x: x[["B", "A"]].values,
+            lambda x: x[["A", "C"]].values.tolist(),
+        ],
+    )
+    def test_setitem_list_missing_columns(self, float_frame, columns, box):
+        # GH 26534
+        result = float_frame.copy()
+        result[columns] = box(float_frame)
```
Comment (WillAyd) [inline pandas/tests/frame/test_indexing.py]: Hmm not really sure what the point of this parametrization is - can you construct test(s) that don't overwrite the existing column data before the assignment? Seems unrelated unless I am missing something

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@howsiwei have been away for a while, can you merge master and i'll look again.

**Comment by jreback** [comment]
@howsiwei can you merge master and i'll have another look

**Comment by WillAyd** [comment]
closing as stale but @howsiwei ping if you'd like to pick this backup

**Comment by howsiwei** [comment]
@WillAyd I'll pick this back up in these few days.

File: pandas/core/frame.py (hunk 14 above)
Comment (howsiwei) [reply pandas/core/frame.py]: Without testing `is_hashable`, at https://github.com/pandas-dev/pandas/blob/3c0cf22909804ae050fdf7609edd13d88d0460aa/pandas/tests/indexing/test_indexing.py#L184 may throw `Buffer has wrong number of dimensions (expected 1, got 2)` or `Array conditional must be same shape as self` instead of the expected errors.

**Comment by WillAyd** [comment]
Sounds good. I don't have the option to reopen though - did you delete the branch on GitHub? If so unfortunately might need to push a new PR

**Comment by howsiwei** [comment]
I suspect it's because I rebase the branch on latest master.

File: pandas/tests/frame/test_indexing.py (hunk 17 above)
Comment (howsiwei) [reply pandas/tests/frame/test_indexing.py]: I parametrize the tests as https://github.com/pandas-dev/pandas/pull/26534#discussion_r306287424 suggest. What do you propose?

**Comment by howsiwei** [comment]
I have created a new PR #29334.
